lib/serve/functions.js

```diff
@@ -0,0 +1,103 @@
+'use strict';
+
+var _ = require('lodash');
+var chalk = require('chalk');
+var path = require('path');
+var RSVP = require('rsvp');
+var getProjectId = require('../getProjectId');
+var logger = require('../logger');
+var utils = require('../utils');
+var parseTriggers = require('../parseTriggers');
+var EmulatorController;
+var controller;
+
+function _pollOperation(service, operation) {
+  return new RSVP.Promise(function(resolve, reject) {
+    var poll = function() {
+      service.operations.get({
+        name: operation[0].name
+      }, function(err, _operation) {
+        if (err) {
+          reject(err);
+        } else if (_operation.done) {
+        // TODO: handle deploy failures
+          resolve(_operation.response.value);
+        } else {
+          setTimeout(poll, 500);
+        }
+      });
+    };
+    poll();
+  });
+}
+
+function _startEmulator(options) {
+  try {
+    EmulatorController = require('@google-cloud/functions-emulator/src/cli/controller');
+  } catch (err) {
+    var msg = err;
+    if (process.version !== 'v6.9.1') {
+      msg = 'Please use Node version v6.9.1, you have ' + process.version + '\n';
+    }
+    utils.logWarning('Cannot start functions emulator. ' + msg);
+    return RSVP.reject();
+  }
+  if (process.version !== 'v6.9.1') {
+    utils.logWarning('The functions emulator works best with Node version v6.9.1, you have ' + process.version + '\n');
+  }
+  var functionsDir = path.join(options.config.projectDir, options.config.get('functions.source'));
+  controller = new EmulatorController({verbose: true});
+  // TODO: pipe logs to console
+  logger.info(chalk.bold('Functions Directory:'), options.config.get('functions.source'), '\n');
+
+  return controller.start().then(function() {
+    return controller.clear();
+  }).then(function() {
+    return parseTriggers(getProjectId(options), options.instance, functionsDir);
+  // TODO: display helpful message when there are errors from trigger parsing
+  }).then(function(triggers) {
+    var promises = _.map(triggers, function(trigger) {
+      if (trigger.httpsTrigger) {
+        return controller.deploy(trigger.name, {
```

Comment:
> The fact that you have to asynchronously "deploy" functions to the emulator seems tough.

Not sure how it could be done otherwise. The Emulator emulates the Google Cloud Functions API, so `Controller#deploy()` is making an RPC call to the emulated API (the same RPC call, in fact, that `firebase deploy` makes when deploying the functions for real). The API uses Long Running Operations, so not only do we have the asynchronous nature of the an RPC call, but if you _really_ want to make sure the functions have been successfully deployed then you have to poll the operations until they've all been marked as `done` with a `response`.

The Firebase SDK could live on the edge and simply ignore the operation objects returned by `Controller#deploy()` and just assume that all the functions have deployed. They'll all be deploying in parallel, and it only takes a couple seconds. So `firebase serve` could immediately report as "ready", and but the time the user actually takes any action to invoke a function the Emulator will have started up and the functions will have deployed (most likely).